### PR TITLE
some cleanups of flexattention full block support

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130605
* #130626

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_higher_order_ops/flex_attention.py [hunk 1]
```diff
@@ -706,26 +686,14 @@ def sdpa_dense_backward(
...
-    )
-    joint_score_mod = torch.vmap(
-        joint_score_mod,
-        in_dims=(0, None, 0, None, None, 0) + in_dim_buffers,
-        out_dims=out_dims,
-    )
-    joint_score_mod = torch.vmap(
-        joint_score_mod,
-        in_dims=(0, 0, None, None, None, 0) + in_dim_buffers,
+        prefix=(0,),
```
Comment (drisspg) [inline torch/_higher_order_ops/flex_attention.py:695]: nit: maybe a comment on why we need this

File: torch/_inductor/kernel/flex_attention.py [hunk 2]
```diff
@@ -188,100 +192,106 @@ def convert_output_node_to_buffer(output):
     l_i = tl.zeros([BLOCK_M], dtype=tl.float32)
     acc = tl.zeros([BLOCK_M, BLOCK_DMODEL], dtype=tl.float32)
 
-    # ~~~~~~~~~~~~~~ fully unmasked blocks ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-    # FULL_KV_IDX and FULL_KV_NUM_BLKS are always contiguous.
+    offs_m = q_start * BLOCK_M + tl.arange(0, BLOCK_M)
+
+    # FULL_KV_IDX and KV_NUM_BLKS are always contiguous.
     sparse_hz_offset = sparse_idx_z * SPARSE_H + sparse_idx_h
```
Comment (drisspg) [inline torch/_inductor/kernel/flex_attention.py:198]: Noob Q: Lets say mask fn was const over B and H, could we make a  sparse_kv_idx non-contiguous by broadcasting  the B, H dims

Comment (drisspg) [reply torch/_inductor/kernel/flex_attention.py:198]: also feels safer to just use the stride info

File: torch/nn/attention/_flex_attention.py [hunk 3]
```diff
@@ -65,59 +111,57 @@ def _no_mask(
...
-    partial_q_indices: torch.Tensor
+    kv_num_blocks: torch.Tensor
+    kv_indices: torch.Tensor
+    q_num_blocks: torch.Tensor
+    q_indices: torch.Tensor
+    full_kv_num_blocks: Optional[torch.Tensor]
+    full_kv_indices: Optional[torch.Tensor]
+    full_q_num_blocks: Optional[torch.Tensor]
+    full_q_indices: Optional[torch.Tensor]
     KV_BLOCK_SIZE: int
```
Comment (Chillee) [inline torch/nn/attention/_flex_attention.py:122]: I renamed the "full vs partial" blocks.

# Before:
 **no mask fn**
full_kv_num_blocks = all the blocks computed
partial_kv_num_blocks = placeholder empty value with all zeros
** with mask fn**
full_kv_num_blocks = only the blocks that are completely full
partial_kv_num_blocks = blocks that aren't full and aren't empty.

# After:
 **no mask fn**
kv_num_blocks = all the blocks computed
full_kv_num_blocks = None
** with mask fn**
kv_num_blocks = the blocks that are partially computed
full_kv_num_blocks = only the blocks that we know are completely full

This makes more sense to me. Basically, `kv_num_blocks` are blocks that we don't know anything special about. So we always need to apply both the `score_mod` and the `mask_mod`. If passed a `mask_mod` to BlockMask and given `full_kv_num_blocks`, then we know have some blocks that we know are "special" and don't have masking applied. So we only need to apply `score_mod`.

**Pull request closed without merging**